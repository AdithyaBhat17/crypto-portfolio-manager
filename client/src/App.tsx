import { Suspense } from "react";
import Layout from "./components/Layout";
import Loading from "./components/Loading";
import Navbar from "./components/Navbar";
import TopCurrencies from "./components/TopCurrencies";
import Footer from "./components/Footer";
import { CurrencyProvider } from "./context/exchange";

function App() {
  return (
    <Layout>
      <CurrencyProvider>
        <Navbar />
        <Suspense fallback={<Loading />}>
          <TopCurrencies />
        </Suspense>
      </CurrencyProvider>
      <Footer />
    </Layout>
  );
}

export default App;
