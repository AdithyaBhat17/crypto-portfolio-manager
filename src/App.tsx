import { Suspense } from "react";
import Layout from "./components/common/Layout";
import Loading from "./components/common/Loading";
import Navbar from "./components/common/Navbar";
import PortfolioManager from "./components/portfolio-manager";
import Footer from "./components/common/Footer";
import { CurrencyProvider } from "./context/exchange";
import ErrorBoundary from "./components/errors/ErrorBoundary";

function App() {
  return (
    <Layout>
      <CurrencyProvider>
        <Navbar />
        <ErrorBoundary>
          <Suspense fallback={<Loading />}>
            <PortfolioManager />
          </Suspense>
        </ErrorBoundary>
      </CurrencyProvider>
      <Footer />
    </Layout>
  );
}

export default App;
