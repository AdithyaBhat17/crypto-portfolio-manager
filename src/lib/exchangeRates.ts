export const CURRENCY_KEY = "MUDREX_USER_CURRENCY";

export const SUPPORTED_CURRENCIES = {
  INR: { symbol: "â‚¹", rate: 73.61 },
  USD: { symbol: "$", rate: 1 },
};

export type Currency = keyof typeof SUPPORTED_CURRENCIES;

function isValidCurrency(currency: any): asserts currency is Currency {
  if (!(currency in SUPPORTED_CURRENCIES))
    throw new TypeError(`${currency} is not supported`);
  return;
}

export function getUserCurrency(): Currency {
  const persistedCurrency = localStorage.getItem(CURRENCY_KEY);
  isValidCurrency(persistedCurrency);
  return persistedCurrency || "INR";
}

export function setUserCurrency(currency: Currency): void {
  localStorage.setItem(CURRENCY_KEY, currency);
}
