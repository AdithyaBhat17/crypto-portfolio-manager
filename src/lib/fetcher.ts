type Method = "GET" | "POST" | "PUT" | "DELETE";

export const URL = process.env.REACT_APP_API_URL || "http://localhost:8080";

export async function fetcher(
  endpoint: string,
  method: Method = "GET",
  body?: string
) {
  return fetch(URL + endpoint, {
    method,
    headers: {
      "Content-Type": "application/json",
      "Accept-Encoding": "deflate,gzip",
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  })
    .then((response) => response.json())
    .catch((e) => {
      throw new Error(e);
    });
}
