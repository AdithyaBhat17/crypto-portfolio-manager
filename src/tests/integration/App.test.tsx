import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import App from "../../App";
import { data } from "../../lib/topCryptoCurrencies__mock__";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { URL } from "../../lib/fetcher";
import { SUPPORTED_CURRENCIES } from "../../lib/exchangeRates";

const server = setupServer(
  rest.get(URL + "/latest", (req, res, ctx) => {
    return res(
      ctx.json({
        data,
      })
    );
  }),
  rest.get(
    "https://assets4.lottiefiles.com/private_files/lf30_jspeqlsz.json",
    (req, res, ctx) => {
      return res(ctx.body("fetched loading spinner"));
    }
  )
);

beforeAll(() => {
  // Establish requests interception layer before all tests.
  server.listen({
    onUnhandledRequest: "error",
  });
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  // Clean up after all tests are done, preventing this
  // interception layer from affecting irrelevant tests.
  server.close();
});

function renderApp() {
  const renderedApp = render(<App />);

  const loadingAnimation = screen.queryByLabelText("loading");

  return {
    ...renderedApp,
    loadingAnimation,
  };
}

test("Renders the Portfolio Manager", async () => {
  const { container, loadingAnimation } = renderApp();

  expect(loadingAnimation).toBeInTheDocument();

  await waitFor(() => expect(loadingAnimation).not.toBeInTheDocument());

  expect(container.innerHTML).toMatchSnapshot();
});

test("User can update preferred exchange rate", async () => {
  localStorage.setItem("holdings", '{"1": "1"}');

  const { loadingAnimation } = renderApp();

  await waitFor(() => expect(loadingAnimation).not.toBeInTheDocument());

  expect(
    (await screen.findByTestId("portfolio value-current-value")).textContent
  ).toBe(
    "₹" +
      (
        data[0].quote.USD.price * SUPPORTED_CURRENCIES["INR"].rate
      ).toLocaleString()
  );

  const selectExchangeRate = screen.getByRole("combobox", {
    name: /exchange-currency/iu,
  });

  fireEvent.change(selectExchangeRate, {
    target: { value: "USD" },
  });

  expect(
    (await screen.findByTestId("portfolio value-current-value")).textContent
  ).toBe("$" + data[0].quote.USD.price.toLocaleString());

  localStorage.clear();
});

test("User can update their holdings", async () => {
  const { loadingAnimation } = renderApp();
  await waitFor(() => expect(loadingAnimation).not.toBeInTheDocument());

  fireEvent.click(
    screen.getAllByRole("button", {
      name: /update holdings/i,
    })[0]
  );

  fireEvent.change(screen.getByRole("spinbutton"), {
    target: { value: 5 },
  });
  fireEvent.blur(screen.getByRole("spinbutton"));

  expect(
    (await screen.findByTestId("portfolio value-current-value")).textContent
  ).toBe(
    "₹" +
      (
        data[0].quote.USD.price *
        SUPPORTED_CURRENCIES["INR"].rate *
        5
      ).toLocaleString()
  );

  // Let's test if user can reset their holding.
  fireEvent.click(
    screen.getAllByRole("button", {
      name: /update holdings/i,
    })[0]
  );

  fireEvent.change(screen.getByRole("spinbutton"), {
    target: { value: 0 },
  });
  fireEvent.blur(screen.getByRole("spinbutton"));

  expect(
    screen.queryByTestId("portfolio value-current-value")
  ).not.toBeInTheDocument();
});
