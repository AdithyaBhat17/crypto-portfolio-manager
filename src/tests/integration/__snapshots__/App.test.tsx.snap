// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the Portfolio Manager 1`] = `"<div class=\\"css-1mpq5ag\\"><header class=\\"css-yq6yn0\\"><h2 class=\\"chakra-heading css-1txeb3g\\">Mudrex </h2><ul role=\\"list\\" class=\\"css-1oxum4v\\"><div class=\\"chakra-stack css-3o20c4\\"><li class=\\"css-0\\"><a href=\\"http://mudrex.com\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">About</a></li><li class=\\"css-0\\"><a href=\\"mailto:support@mudrex.com\\">Contact</a></li><li class=\\"css-0\\"><div class=\\"chakra-select__wrapper css-1ilqd55\\"><select aria-label=\\"exchange-currency\\" class=\\"chakra-select css-ab5buk\\"><option value=\\"INR\\" selected=\\"\\">â‚¹ INR</option><option value=\\"USD\\">$ USD</option></select><div class=\\"chakra-select__icon-wrapper css-162mkon\\"><svg viewBox=\\"0 0 24 24\\" role=\\"presentation\\" class=\\"chakra-select__icon\\" focusable=\\"false\\" aria-hidden=\\"true\\" style=\\"width: 1em; height: 1em; color: currentColor;\\"><path fill=\\"currentColor\\" d=\\"M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z\\"></path></svg></div></div></li></div></ul></header><div class=\\"css-1s479nb\\"><table role=\\"table\\" class=\\"chakra-table css-0\\"><thead class=\\"css-0\\"><tr role=\\"row\\" class=\\"css-0\\"><th class=\\"css-xi606m\\">#</th><th class=\\"css-vc5bon\\">Name</th><th class=\\"css-0\\">Symbol</th><th data-is-numeric=\\"true\\" class=\\"css-0\\">Price (INR)</th><th data-is-numeric=\\"true\\" class=\\"css-0\\"><div class=\\"css-192qrng\\"><span class=\\"chakra-text css-ym5um8\\">Circulating supply</span> </div></th><th data-is-numeric=\\"true\\" class=\\"css-0\\"><div class=\\"css-192qrng\\"><span class=\\"chakra-text css-ym5um8\\">Your holdings</span> </div></th></tr></thead><tbody class=\\"css-0\\"><tr role=\\"row\\" class=\\"css-axuxj7\\"><td role=\\"gridcell\\" class=\\"css-xi606m\\">5</td><td role=\\"gridcell\\" class=\\"css-11vo1xj\\">Bitcoin</td><td role=\\"gridcell\\" class=\\"css-0\\">BTC</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\">â‚¹683,389.351</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\">16,950,100</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\"><span class=\\"css-192qrng\\"><span>-</span><button type=\\"button\\" class=\\"chakra-button css-wpkkii\\" aria-label=\\"Update holdings\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><g fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-width=\\"2\\"><path d=\\"M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7\\"></path><path d=\\"M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z\\"></path></g></svg></button></span></td></tr><tr role=\\"row\\" class=\\"css-axuxj7\\"><td role=\\"gridcell\\" class=\\"css-xi606m\\">-</td><td role=\\"gridcell\\" class=\\"css-11vo1xj\\">Ethereum</td><td role=\\"gridcell\\" class=\\"css-0\\">ETH</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\">â‚¹94,509.351</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\">16,950,100</td><td role=\\"gridcell\\" data-is-numeric=\\"true\\" class=\\"css-0\\"><span class=\\"css-192qrng\\"><span>-</span><button type=\\"button\\" class=\\"chakra-button css-wpkkii\\" aria-label=\\"Update holdings\\"><svg viewBox=\\"0 0 24 24\\" focusable=\\"false\\" class=\\"chakra-icon css-onkibi\\" aria-hidden=\\"true\\"><g fill=\\"none\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\" stroke-width=\\"2\\"><path d=\\"M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7\\"></path><path d=\\"M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z\\"></path></g></svg></button></span></td></tr></tbody></table></div><p class=\\"chakra-text css-1iu1318\\">To the moon ðŸš€</p></div>"`;
